Batteries/Control/LawfulMonadState.lean

```diff
@@ -0,0 +1,225 @@
+/-
+Copyright (c) 2025 Devon Tuma. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Devon Tuma, Quang Dao
+-/
+
+/-!
+# Laws for Monads with State
+
+This file defines a typeclass for `MonadStateOf` with compatible `get` and `set` operations.
+
+Note that we use `MonadStateOf` over `MonadState` as the first induces the second,
+but we phrase things using `MonadStateOf.set` and `MonadState.get` as those are the
+versions that are available at the top level namespace.
+-/
+
+/-- The namespaced `MonadStateOf.get` is equal to the `MonadState` provided `get`. -/
+@[simp] theorem monadStateOf_get_eq_get [MonadStateOf σ m] :
+    (MonadStateOf.get : m σ) = get := rfl
+
+/-- The namespaced `MonadStateOf.modifyGet` is equal to the `MonadState` provided `modifyGet`. -/
+@[simp] theorem monadStateOf_modifyGet_eq_modifyGet [MonadStateOf σ m]
+    (f : σ → α × σ) : (MonadStateOf.modifyGet f : m α) = modifyGet f := rfl
+
+/-- Class for well behaved state monads, extending the base `MonadState` type.
+Requires that `modifyGet` is equal to the same definition with only `get` and `set`,
+that `get` is idempotent if the result isn't used, and that `get` after `set` returns
+exactly the value that was previously `set`. -/
+class LawfulMonadStateOf (σ : Type _) (m : Type _ → Type _) [Monad m] [MonadStateOf σ m] where
```

Comment:
Is there a utility for a `LawfulMonadStateOf` that is not a `LawfulMonad` too?

Reply:
Not that I'm aware of, just instinct to minimize assumptions in things. I can add that in, I'm not sure if `extends` might lead to diamond issues with `EStateM` getting a similar instance for lawful error behavior.